### DISCUSSION TITLE
Tuples as pairs

### DISCUSSION THREAD

**Discussion by andfenastari** [gleam-lang/gleam#1474] [Ideas & suggestions]
# Motivation

A problem I often see on languages with rust-like type systems is the proliferation of "numbered functions", that is, variants of functions operating on heterogeneous collections based on their size, e.g. gleam/dynamics `tuple2`, `tuple3`, `tuple4`, etc.
Functions like these are such a common pattern (I wouldn't be surprised if the iterator module received higher arity versions of the `zip` function at some point in the future) as always involve such amounts of hackery that I think the language ought to support some solution for it, if not the one hereby presented.

# Proposal

Introduce a new primitive type `pair(a, b)` and make tuple types syntactic sugar for it, so that `#(a, b)` means `pair(a, c)`,
`#(a, b, c)` means `pair(pair(a, b), c)`, etc. This way, the `tuple{N}` functions can be replaced by a single `fn tuple(first: Decoder(a), second: Decoder(b)) -> Decoder(pair(a, b))`, which then could be chained with the `|>` operator.
Another example using iterators:
```gleam
// An iterator listing all of the Pythagorean triples with side length <= to n.
fn pyth(to n: Int) -> Iterator(#(Int, Int, Int)) {
   let ns = iterator.range(n)
   ns
   |> iterator.zip(ns)
   |> iterator.zip(ns)
   |> iterator.filter(fn(#(a, b, c)) {
     a*a == b*b + c*c
   })
}
```

# Further thoughts
Another feature that would combine nicely with this one, but might be too complicated for the language, is the ability to "spread" tuples as arguments to functions, allowing the `decode{N}` functions to be replaced by a combination of `fn decode(constructor: fn(..args) -> t, decoder: Decoder(args)) -> Decoder(t)` and `tuple`, as in:
```
type Book {
  Book(title: String, author: String, timestamp: Int)
 }
 
fn decode_book(value: Dynamic) -> Result(Book, List(DecodeError)) {
  decode(
    Book,
    field("title", string)
    |> tuple(field("author", string))
    |> tuple(field("timestamp", integer))
  )
}
```

**Comment by lpil** [comment]
@andfenastari Hello! Thanks for your suggestion.
I agree, the ergonomics of tuples could certainly be improved. These `thingN` functions are not the prettiest of APIs, and they are annoying to implement.

One big restriction we have is FFI, we want to remain compatible with Erlang data structures. What Erlang code would we generated for these expresssions?

```rust
pair(1, 2)
pair(1, pair(2, 3))
pair(1, pair(2, pair(3, 4)))
```

If `#(...)` is sugar for `pair` how would we represent `#()` and `#(1)`?
